Add support for BallTree, BinaryTree

BenjaminBossan opened the pull request:
Those are used by `KNeighbors{Classifier,Regressor}` but were not caught so far because of the default argument.

Unfortunately, we need some special treatment here because the two cython classes, as well as `DistanceMetric`, which is used under the hood, return tuples from `__getstate__` and thus the normal `reduce_get_state` and `reduce_get_instance` methods fail.

LMK if you have a better name than `reduce_based_on_tuple_get_{state,instance}`.

Ready for review @adrinjalali

BenjaminBossan committed:
Add support for BallTree, BinaryTree

Those are used by KNeighbors{Classifier,Regressor} but were not caught
so far because of the default argument.

Unfortunately, we need some special treatment here because the two
cython classes, as well as DistanceMetric, which is used under the hood,
return tuples from __getstate__ and thus the normal reduce_get_state and
reduce_get_instance methods fail.

adrinjalali commented on skops/io/_sklearn.py:
```diff
@@ -171,14 +176,52 @@ def reduce_get_instance(state, src, constructor):
+
+
+def reduce_based_on_tuple_get_state(obj, dst):
+    # For more details, take a look at reduce_get_state. This extra function is
+    # needed because __getstate__ of certain sklearn classes returns a tuple
+    # instead of a dict, which we thus need to treat differently.
```
why not change reduce so that it can handle a tuple/list?

BenjaminBossan replied:
I was thinking about that but was afraid that the function, which is already quite complex, would become too confusing. If you think it's okay, I can work it in.

adrinjalali replied:
No I think that function deserves to handle this complexity.

BenjaminBossan replied:
Is this a "no, don't merge the functions" or a "no, don't keep them separate"? :D

adrinjalali replied:
lolol, very fair. As in: I think the basic function should handle tuples, so please merge them together and have tuple handling in `reduce_*`.

adrinjalali commented on skops/io/_sklearn.py:
```diff
@@ -13,6 +13,9 @@
     SquaredLoss,
 )
+from sklearn.metrics import DistanceMetric
+from sklearn.metrics._dist_metrics import METRIC_MAPPING
```
wouldn't maybe a user be able to modify this and hence allow things which shouldn't be allowed?

BenjaminBossan replied:
So would you suggest to hard-code the metrics instead?

adrinjalali replied:
yeah, like I did for losses.

BenjaminBossan committed:
Merge branch 'adrinjalali-persist' into perist-binary-trees

BenjaminBossan committed:
Hard-code allowed distance metrics

BenjaminBossan commented:
@adrinjalali I pushed the change of hard-coding the distance metrics.

BenjaminBossan committed:
Merge branch 'adrinjalali-persist' into perist-binary-trees

BenjaminBossan committed:
Refactor reduce_get_state/instance

Instead of having separate functions for tuples, now one function deals
with both tuples and dicts.

BenjaminBossan commented:
@adrinjalali I addressed your points, please review again.

adrinjalali commented on skops/io/_sklearn.py:
```diff
@@ -144,10 +189,19 @@ def reduce_get_state(obj, dst):
-        content[key] = _get_state(value, dst)
+    # the return value from __getstate__ (and thus indirectly __reduce__) is
+    # usually a dict but some classes return tuples
+    if isinstance(attrs, tuple):
+        content["args"] = {
+            "__class__": obj.__class__.__name__,
+            "__module__": get_module(type(obj)),
+            "__reduce__": {"state": _get_state(attrs, dst)},
+        }
+    else:
+        for key, value in attrs.items():
+            if isinstance(getattr(type(obj), key, None), property):
+                continue
+            content[key] = _get_state(value, dst)
```
can't we simply do:

```python
content["attrs"] = _get_state(attrs, dst)
```

and in the `reduce_get_instance`:

```python
args = _get_instance(state["content"]["attrs"], src)

if isinstance(args, tuple) and not hasattr(instance, "__setstate__"):
    raise

if hasattr(instance, "__setstate__"):
    instance.__setstate__(attrs)
else:
    instance.__dict__.update(attrs)
```

adrinjalali reviewed (commented):
WDYT

BenjaminBossan committed:
Merge branch 'adrinjalali-persist' into perist-binary-trees

BenjaminBossan committed:
Big simplification of code

- Big refactor of reduce_get_state/instance
- Thus, no special treatment required for BinaryTree and DistanceMetric
  anymore
- Moreover, some simplification of object_get_state

BenjaminBossan committed:
Merge branch 'adrinjalali-persist' into perist-binary-trees

BenjaminBossan commented:
@adrinjalali Your suggestion wasn't quite enough, but it pointed me towards a huge simplification :+1: Please take a look again.

adrinjalali reviewed (approved):
Very nice!
